upload_vehicle_data.py

```diff
@@ -0,0 +1,729 @@
+            logger.error(f'Failed to create directory {directory_path}: {e}')
+            raise
+
+
+def process_directory(
+    logger,
+    remote_user,
+    remote_ip,
+    remote_directory,
+    cloud_upload_directory,
+    config,
+    base_remote_directory,
+    bandwidth_mbps,
+):
+    """Process each directory."""
+    # Create the remote temporary directory
+    create_remote_temp_directory(
+        logger, remote_user, remote_ip, remote_directory
+    )
+
+    try:
+        # Find and copy the metadata.yaml file
+        metadata_path = find_metadata_file(
+            logger, remote_user, remote_ip, remote_directory
+        )
+        if metadata_path is None:
+            logger.error(
+                f'metadata.yaml file not found in {remote_directory}. '
+                f'Skipping.'
+            )
+        relative_metadata_path = os.path.relpath(
+            metadata_path, start=base_remote_directory
+        )
+
+        local_metadata_path = os.path.join(
+            cloud_upload_directory, relative_metadata_path
+        )
+
+        check_and_create_local_directory(
+            logger, os.path.dirname(local_metadata_path)
+        )
+
+        if not copy_metadata_file(
+            logger,
+            remote_user,
+            remote_ip,
+            metadata_path,
+            cloud_upload_directory,
+            base_remote_directory,
+        ):
+            logger.error(
+                f'Failed to copy metadata.yaml from '
+                f'{remote_directory}. Skipping.'
+            )
+
+        # Read the metadata.yaml file
+
+        metadata = read_metadata(logger, local_metadata_path)
+        expected_bags = metadata.get('rosbag2_bagfile_information', {}).get(
+            'relative_file_paths', None
+        )
+
+        # Get the list of rosbags from the remote machine
+        rosbag_list = get_remote_rosbags_list(
+            logger, remote_user, remote_ip, remote_directory
+        )
+
+        if len(rosbag_list) != len(expected_bags):
+            logger.error(
+                'The number of rosbags does not match the metadata. Skipping.'
+            )
+            return
+
+        rosbag_sizes = [
+            get_remote_file_size(logger, remote_user, remote_ip, rosbag)
+            for rosbag in rosbag_list
+        ]
+        total_size_gb = sum(rosbag_sizes)
+        estimated_time = get_estimated_upload_time(
+            total_size_gb, bandwidth_mbps, rosbag_sizes
+        )
+
+        if estimated_time < 1:
+            estimated_time_str = f'{estimated_time * 60:.2f} minutes'
+        else:
+            estimated_time_str = f'{estimated_time:.2f} hours'
+
+        logger.info(
+            f'Found {len(rosbag_list)} files to upload with total size '
+            f'{total_size_gb:.2f} GB from {remote_directory}.'
+        )
+        logger.info(
+            f'Estimated total time (including compression) '
+            f'is at least: {estimated_time_str}.'
+        )

```

Comment:
I do not think we need to call `get_remote_file_size()` and `get_remote_file_size()` function here again as we already called them inside `main` `for` loop. This means we already get all `.mcap` files paths and its sizes so there is no need to compute this again. This can slow down our scirpt if each didrectoy contain more than 2 rosbags. Maybe we can implement a directionary that stores all of these, then this `process_directory()` function can read from it based on the target subdirectory.

Reply:
Maybe worth for you to measure how long these functions take. Do not push this time measurement just shared with us. Then we can see if is worth or not implement the optimisation I explained above.

Reply:
Changed